arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,599 @@
+	movl %eax, %cr4
+	clts
+
+	movl $pml4, %eax		/* load page base */
+	movl %eax, %cr3
+
+	movl $X86_EFER_MSR, %ecx	/* enable long mode */
+	rdmsr
+	orl $X86_EFER_MSR_LME, %eax
+	wrmsr
+
+	movl %cr0, %eax			/* enable paging */
+	orl $CR0_PG, %eax
+	movl %eax, %cr0
+
+	/* jump into long mode, reload the segment registers (again). */
+
+	jmpl $X86_KERNEL_CS_64, $1f
+.code64
+1:	movl $X86_KERNEL_DS_64, %eax
+	movw %ax, %ds
+	movw %ax, %es
+	movw %ax, %ss
+	movw %ax, %fs
+	movw %ax, %gs
+
+	movl $X86_KERNEL_TSS, %eax
+	ltr %ax
+
+	cld
+	xorl %eax, %eax
+	movq $__bss_start, %rdi
+	movq $__bss_num_qwords, %rcx
+	rep stosq
+
+#ifdef CONFIG_X86_MULTIBOOT_INFO
+	/* If EBX isn't 0, then we have a valid multiboot info struct. */
+
+	testl %ebx, %ebx
+	jz 1f
+	movl %ebx, %esi
+
+	movl $__X86_MULTIBOOT_INFO_SIZEOF, %ecx
+	movq $x86_multiboot_info, %rdi
+	rep
+	movsb
+1:
+#endif /* CONFIG_X86_MULTIBOOT_INFO */
+
+	/*
+	 * set up SSE in case something uses the floating-point unit during
+	 * early initialization (either directly, or if GCC gets clever)
+	 */
+
+	xorl %edi, %edi
+	call x86_sse_init
+
+	/* don't replace CALL with JMP; honor the ABI stack alignment! */
+
+	call z_cstart
+
+stop:	jmp stop
+
+/*
+ * void x86_sse_init(struct k_thread *thread);
+ *
+ * Initialize floating-point state to something sane. If 'thread' is
+ * not NULL, then the resulting FP state is saved to thread->arch.sse.
+ */
+
+.global x86_sse_init
+x86_sse_init:
+	fninit
+	ldmxcsr mxcsr
+	testq %rdi, %rdi
+	jz 1f
+	fxsave _thread_offset_to_sse(%rdi)
+1:	retq
+
+mxcsr:	.long X86_MXCSR_SANE
+
+/*
+ * FIXME: The multiboot header is identical (for obvious reasons) to the
+ * version in ia32/crt0.S. They should be refactored into a common file.
+ */
+
+#ifdef CONFIG_X86_MULTIBOOT
+	.align 4
+	.long X86_MULTIBOOT_HEADER_MAGIC
+	.long X86_MULTIBOOT_HEADER_FLAGS
+	.long -(X86_MULTIBOOT_HEADER_MAGIC + X86_MULTIBOOT_HEADER_FLAGS)
+#ifdef CONFIG_X86_MULTIBOOT_FRAMEBUF
+	.fill 5,4,0				/* (unused exec layout) */
+	.long 0					/* linear graphics mode */
+	.long CONFIG_X86_MULTIBOOT_FRAMEBUF_X	/* width */
+	.long CONFIG_X86_MULTIBOOT_FRAMEBUF_Y	/* height */
+	.long 32				/* depth */
+#endif /* CONFIG_X86_MULTIBOOT_FRAMEBUF */
+#endif
+
+/*
+ * XXX: describe __swap, __resume, stacks
+ */
+
+.globl _k_neg_eagain	/* from errno.c: int _k_neg_eagain = -EAGAIN; */
+
+.globl __swap
+__swap:
+	movq $_kernel, %rsi
+	movq _kernel_offset_to_current(%rsi), %rsi
+
+	orb $_THREAD_SWAPPED, _thread_offset_to_thread_state(%rsi)
+
+	movl _k_neg_eagain, %eax
+	movl %eax, _thread_offset_to_rax(%rsi)
+	popq %rax
+	movq %rax, _thread_offset_to_rip(%rsi)
+	movq %rsp, _thread_offset_to_rsp(%rsi)
+	movl %edi, %edi /* N.B.: zero extend */
+	movq %rdi, _thread_offset_to_rflags(%rsi)
+	movq %rbx, _thread_offset_to_rbx(%rsi)
+	movq %rbp, _thread_offset_to_rbp(%rsi)
+	movq %r12, _thread_offset_to_r12(%rsi)
+	movq %r13, _thread_offset_to_r13(%rsi)
+	movq %r14, _thread_offset_to_r14(%rsi)
+	movq %r15, _thread_offset_to_r15(%rsi)
```

Comment:
By my count this makes four separate context save blocks.  The original code made do with just one, and z_arch_swich() was just a 6-instruction inline assembly block which used the embedded CALL instruction to populate the stack frame that would be used to restore.

Also this has the problem mentioned elsewhere that doing a stack save with MOV vs. PUSH involves trippling your icache footprint.

Reply:
> By my count this makes four separate context save blocks. The original code made do with just one, and z_arch_swich() was just a 6-instruction inline assembly block which used the embedded CALL instruction to populate the stack frame that would be used to restore.

There are **three** context save (and restore) scenarios, not four.  (The exception "context save" is not a context save at all.)

Comparing this to the "original code" somehow implies that I derived this implementation from yours.  I did not.  This implementation is different and has different design constraints, and as a result I made different design choices.  Things are complicated here than your implementation because I don't use `-fno-red-zone` and have to use the IST mechanism.  You're comparing apples and oranges.

> Also this has the problem mentioned elsewhere that doing a stack save with MOV vs. PUSH involves trippling your icache footprint.

See previous comments.  The performance implications of MOV vs PUSH are not as cut-and-dried as you make them out to be, and I hardly think a few hundred more bytes in a .. what, 32K i-cache? .. is worth getting excited about.

Reply:
And really, here, w/r/t icache: we're not even talking about a 'few' hundred more bytes.  Probably less than a hundred, total, across the entirely of the locore, with the MOVs vs. PUSHes. Even this could be trimmed down if the arch register struct weren't the last member of `k_thread`, though there appears to be (or at least, was at one time) a reason this member is at the end. But again, I don't the icache footprint is a worthy of much attention here.